### PR TITLE
Heatmap + Batch Processing

### PR DESCRIPTION
**Pull request by trbritt**
This PR is for a miscallaneous assortment of new nodes to be included.

Firstly, I've revamped the heatmap visualizer to now include a 'histogram' option. With this option, a secondary axis will appear to the right of the render with the histogram counts of the image, for easy inspection of the spectral content. Second, a slider appears in the bottom of the Plotly figure to dynamically adjust the `zmax` value to tune the contrast in real time.

output_histogram

Secondly, I've created a batch processing node. The idea of this node to be able to glob match files in a specific directory, iterate over these files, and perform the processing on them one by one. It is also equipped to handle dynamically increasing number of files in the directory with the `refresh` parameter. For example, if a separate process is continuing to add new images to the folder, then the node will add the new files to the queue to be processed.

app

Given the dynamically changing number of files, the application uses an infinite loop, checks when the processing queue is empty, and then breaks the loop. Of course to break the loop required the development of the `BREAK` node for this purpose, put in the `LOOP_TOOLS` category.

app

None of these have unit tests built in, but they all have example apps and renders.

### CODE REVIEW

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
+    JobResultBuilder,
+)
 
 
 @flojoy
 def FEEDBACK(
     referred_node: NodeReference,
     default: Optional[DataContainer] = None,
-) -> Optional[DataContainer]:
+) -> Any:
```
Comment (itsjoeoui) [inline GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:15]: @JeffDotPng Not sure if we are okay with changing this to `Any` ?

Comment (JeffDotPng) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:15]: Yes this should stay as `Optional[DataContainer]`

Comment (trbritt) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py]: I've pushed the change just now!

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py [hunk 2]
```diff
@@ -20,4 +26,9 @@ def FEEDBACK(
     if result:
         return result
     else:
-        return default
+        return (
```
Comment (JeffDotPng) [inline GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:29]: Why was this changed?

File: LOADERS/LOCAL_FILE_SYSTEM/BATCH_PROCESSOR/BATCH_PROCESSOR.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        files = get_fnames(directory_path, pattern if pattern else "*")
+        SmallMemory().write_to_memory(
+            node_id,
+            memory_key,
+            {
+                "node_id": node_id,
+                "current_iteration": curr_iter,
+                "files": files,
+                "original_files": files,
+                "n_files": int(len(files)),
```
Comment (JeffDotPng) [inline LOADERS/LOCAL_FILE_SYSTEM/BATCH_PROCESSOR/BATCH_PROCESSOR.py:80]: Unnecessary int cast?

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py (hunk 1 above)
Comment (itsjoeoui) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py]: Thanks!

File: LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py [hunk 4]
```diff
@@ -29,8 +29,7 @@ def CONDITIONAL(
     # y_of_x = x.y
     # y_of_y = y.y
 
-    bool_ = compare_values(x.c, y.c, operator_type)
-
+    bool_ = compare_values(x.c[0], y.c[0], operator_type)
```
Comment (JeffDotPng) [inline LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py:32]: Why indexing? If this is needed, something is wrong.. :thinking:

File: LOGIC_GATES/LOOP_TOOLS/BREAK/BREAK.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+from typing import Optional, Any
+
+memory_key = "loop-info"
+
+
+@flojoy
+def BREAK(
+    referred_node: NodeReference,
+    default: Optional[DataContainer] = None,
+) -> DataContainer:
```
Comment (JeffDotPng) [inline LOGIC_GATES/LOOP_TOOLS/BREAK/BREAK.py:11]: Should this even return anything? I think it actually should just return `None` to prevent users from connecting anything to the output, a break node should always be a leaf in the DAG

File: LOGIC_GATES/LOOPS/LOOP/LOOP.py [hunk 6]
```diff
@@ -70,12 +70,23 @@ def LOOP(
 
     node_id = default_params.node_id
 
+    loop_data: LoopData = load_loop_data(node_id, num_loops)
+
+    # given the addition of the break node, it is possible that
+    # another node can write to the data of this loop. we have to
+    # now check if that's the case, and if so, return
+    if loop_data.get_data().get("is_finished"):
+        return build_result([default] if default else [], True)
```
Comment (JeffDotPng) [inline LOGIC_GATES/LOOPS/LOOP/LOOP.py:79]: Could this potentially break nested loops? This check will always be executed before the `.restart()` so the loop will just exit immediately upon being reached again (?)

File: VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py [hunk 7]
```diff
@@ -56,7 +56,7 @@ def BIG_NUMBER(
         case OrderedPair():
             big_num = default.y[-1]
         case Scalar():
-            big_num = default.c
+            big_num = default.c[-1]
```
Comment (JeffDotPng) [inline VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py:59]: A scalar should just be a float so this shouldn't be necessary (?) I think there might be something wrong with the app

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py (hunk 2 above)
Comment (trbritt) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:29]: This was implemented by @smahmed776 to get around the fact that in some topologies the feedback node didn't direct the execution order to the correct next node

File: LOADERS/LOCAL_FILE_SYSTEM/BATCH_PROCESSOR/BATCH_PROCESSOR.py (hunk 3 above)
Comment (trbritt) [reply LOADERS/LOCAL_FILE_SYSTEM/BATCH_PROCESSOR/BATCH_PROCESSOR.py:80]: indeed, will remove

File: LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py (hunk 4 above)
Comment (trbritt) [reply LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py:32]: Conditional node broke when comparing to Scalar types of different lengths (the usual "the truth value of an array with more than one element is ambiguous" error), despite the fact that the functionality should be defined

File: LOGIC_GATES/LOOPS/LOOP/LOOP.py (hunk 6 above)
Comment (trbritt) [reply LOGIC_GATES/LOOPS/LOOP/LOOP.py:79]: I demoed it on nested loops and it worked well. Soon as loop starts, it writes its data for all nodes to see, and this just catches if other nodes have written to its data in `SmallMemory`. Each loop node has its unique data that the `break` node can see, so you're able to uniquely change a particular instance of a `loop` node and thus break only that loop.

File: LOGIC_GATES/LOOP_TOOLS/BREAK/BREAK.py (hunk 5 above)
Comment (trbritt) [reply LOGIC_GATES/LOOP_TOOLS/BREAK/BREAK.py:11]: I agree this should return nothing!

File: VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py (hunk 7 above)
Comment (trbritt) [reply VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py:59]: It wasnt the app; when I passed a Scalar type to big number, if the scalar type has length >1, then the line `value=int(float(big_num))` fails (only length-1 arrays can be converted to Python scalars), even though that clearly it should be allowed to display anything from a Scalar data container

File: LOGIC_GATES/LOOPS/LOOP/LOOP.py (hunk 6 above)
Comment (JeffDotPng) [reply LOGIC_GATES/LOOPS/LOOP/LOOP.py:79]: I see, the inner loop is able to restart properly on subsequent iterations of the outer loop?

Comment (trbritt) [reply LOGIC_GATES/LOOPS/LOOP/LOOP.py:79]: It is, but what I can do to be extra safe is just add a line to explicitly write `is_finished=False` to the memory of that node to ensure clean restart, probably a good idea

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py (hunk 2 above)
Comment (JeffDotPng) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:29]: That's interesting, I guess because of the None return value? Does it still happen after you added the initial value to the feedback nodes?
Either way,  I didn't catch this before saying to change the return annotation back to `Optional[DataContainer]`, in this case you were right before and it should be `Any` instead

File: LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py (hunk 4 above)
Comment (JeffDotPng) [reply LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py:32]: Scalar shouldn't be an array at all though? It should just be a float

File: VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py (hunk 7 above)
Comment (JeffDotPng) [reply VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py:59]: What was generating a scalar of array-type? This definitely shouldn't happen, it should always be a float

Comment (trbritt) [reply VISUALIZERS/PLOTLY/BIG_NUMBER/BIG_NUMBER.py:59]: It was the constant node, you can avoid the problem by just remembering to set the length to 1, but I made the change to 'user-proof' it and avoid all the potential newcomers that might not have the programming insight to realize you needed to specify length=1

File: LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py (hunk 4 above)
Comment (trbritt) [reply LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py:32]: To illustrate the exact behaviour, without these changes, the loop below will just restart the loop and not follow either of the bodies connected to the conditional node. Only by the changes above will the conditional node actually redirect the workflow to its children branches.

[Screencast from 08-26-2023 01:50:26 PM.webm](https://github.com/flojoy-ai/nodes/assets/53545754/6af8f997-9be0-48aa-8a7e-db1d2e748dee)

File: GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py (hunk 2 above)
Comment (trbritt) [reply GENERATORS/SIMULATIONS/FEEDBACK/FEEDBACK.py:29]: Changed back to Any return type :ok_hand:

File: LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py (hunk 4 above)
Comment (JeffDotPng) [reply LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py:32]: Fixed by this PR: https://github.com/flojoy-ai/python/pull/83

Comment (trbritt) [reply LOGIC_GATES/CONDITIONALS/CONDITIONAL/CONDITIONAL.py]: Done! Will merge the PR now.
